[AArch64][llvm] Armv9.7-A: Add support for GICv5 (FEAT_GCIE)

jthackray opened the pull request:
Add new instruction and system registers that are specified in the
Generic Interrupt Controller Architecture v5 (GICv5) standard,
announced here:

  * https://community.arm.com/arm-community-blogs/b/architectures-and-processors-blog/posts/introducing-gicv5

and documented here:

  * https://developer.arm.com/documentation/109697/2025_09/2025-Architecture-Extensions
  * https://developer.arm.com/documentation/ddi0602/2025-09/

Co-authored-by: Jack Styles <jack.styles@arm.com>

jthackray committed:
[AArch64][llvm] Armv9.7-A: Add support for GICv5 (FEAT_GCIE)

Add new instruction and system registers that are specified in the
Generic Interrupt Controller Architecture v5 (GICv5) standard,
announced here:

  * https://community.arm.com/arm-community-blogs/b/architectures-and-processors-blog/posts/introducing-gicv5

and documented here:

  * https://developer.arm.com/documentation/109697/2025_09/2025-Architecture-Extensions
  * https://developer.arm.com/documentation/ddi0602/2025-09/

Co-authored-by: Jack Styles <jack.styles@arm.com>

jthackray committed:
fixup! [AArch64][llvm] Armv9.7-A: Add support for GICv5 (FEAT_GCIE)

Move FEAT_GCIE from VFAT

CarolineConcatto commented on llvm/test/MC/AArch64/armv9.7a-gcie.s:
```diff
@@ -0,0 +1,985 @@
+// Disassemble encoding and check the re-encoding (-show-encoding) matches.
+// RUN: llvm-mc -triple=aarch64 -show-encoding -mattr=+gcie < %s \
+// RUN:        | sed '/.text/d' | sed 's/.*encoding: //g' \
+// RUN:        | llvm-mc -triple=aarch64 -mattr=+gcie -disassemble -show-encoding \
```
Can we also add a diagnostic test for GIC, GICR and GSB?

jthackray committed:
fixup! [AArch64][llvm] Armv9.7-A: Add support for GICv5 (FEAT_GCIE)

Avoid repetition. PPI and Virtual PPI registers can share code.

jthackray commented on llvm/lib/Target/AArch64/AArch64SystemOperands.td:
```diff
@@ -2480,3 +2480,211 @@ def : MLBI<"ALLE1",    0b100, 0b0111, 0b0000, 0b100, 0>;
+def : GIC<"cdpri",    0b000, 0b1100, 0b0001, 0b010, 1>;
+def : GIC<"cdrcfg",   0b000, 0b1100, 0b0001, 0b101, 1>;
+def : GICR<"cdia",    0b000, 0b1100, 0b0011, 0b000, 1>;
+def : GICR<"cdnmia",  0b000, 0b1100, 0b0011, 0b001, 1>;
```
Sure (I was doing them in the order listed in the spec).

jthackray committed:
fixup! [AArch64][llvm] Armv9.7-A: Add support for GICv5 (FEAT_GCIE)

Optimisations suggested in CR.

CarolineConcatto commented on llvm/lib/Target/AArch64/AArch64Features.td:
```diff
@@ -604,6 +604,9 @@ def FeatureMPAMv2: ExtensionWithMArch<"mpamv2", "MPAMv2", "FEAT_MPAMv2",
   "Enable Virtual Memory Tagging Extension">;
 
+def FeatureGCIE: ExtensionWithMArch<"gcie", "GCIE", "FEAT_GCIE",
+  "Enable Armv9.7-A GICv5 (Generic Interrupt Controller) CPU Interface Extension", [FeatureNMI]>;
```
Should this be removed?

CarolineConcatto commented on llvm/test/MC/AArch64/armv9.7a-gcie.s:
```diff
@@ -0,0 +1,985 @@
+// RUN:        | FileCheck %s --check-prefixes=CHECK-ENCODING,CHECK-INST
+// RUN: not llvm-mc -triple=aarch64 -show-encoding < %s 2>&1 \
+// RUN:        | FileCheck %s --check-prefix=CHECK-ERROR
+// RUN: llvm-mc -triple=aarch64 -filetype=obj -mattr=+all < %s \
```
s/+all/+gcie/g

CarolineConcatto commented on llvm/test/MC/AArch64/armv9.7a-gcie-diagnostics.s:
```diff
@@ -0,0 +1,29 @@
+gsb
+// CHECK-ERROR: error: invalid operand for GSB instruction
+
+gsb ack
```
I dont think we need this here, because you already test in armv9.7a-gcie.s

jthackray replied:
Yep, done.

jthackray committed:
fixup! [AArch64][llvm] Armv9.7-A: Add support for GICv5 (FEAT_GCIE)

Minor test nits, and remove duplicated tests

jthackray committed:
fixup! [AArch64][llvm] Armv9.7-A: Add support for GICv5 (FEAT_GCIE)

Remove `FeatureNMI` as a dependency for FEAT_GCIE. It's in the
GICv5 spec, but we're not going to force it in the compiler.

jthackray committed:
fixup! [AArch64][llvm] Armv9.7-A: Add support for GICv5 (FEAT_GCIE)

Remove Armv9.7-A from GCIE

CarolineConcatto reviewed (approved):
LGTM!

jthackray commented:
### Merge activity

* **Oct 23, 12:32 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/163159).
